### PR TITLE
Feature: print email

### LINKED ISSUE #579: Print stylesheet [$65]
Occasionally a user needs to print an email. CTRL+P will print the page, not the message. Adding a div to the messages-list-item template and some js tied to an action will not address CTRL+P but im not sure if everyone would want to override the browser shortcut anyway.  

There are some tricks in CSS To make the output in the mail-message-container cleaner ( reply attachment and forward buttons for example ) for printing but putting the print button in the message list item would at least prevent that particular control from being printed. 

Alternatively I suppose it would be even better in some cases to pull the message from the mail server again for printing purposes to avoid printing all the extraneous information mentioned above, and in the event someone entered text into the reply but had not sent it yet -- we might not want to print message text that has never been sent. 

Other considerations are whether we want the print preview dialog allowing the user to select a printer or to blindly print to whatever is default.   

Any thoughts on this?  I can probably contribute to this one if its not already in progress or part of NCMail13 but Id like to hear some input from others. 

thx
-d

<bountysource-plugin>

---
There is a **[$65 open bounty](https://www.bountysource.com/issues/50806693-print-stylesheet?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sourenaraya**
Fixes #579

new dependency introduced, not sure if that is ok.
Ctrl+P works.

### CODE REVIEW

**Comment by sourenaraya** [comment]
any update on review?

File: package.json [hunk 1]
```diff
@@ -31,6 +31,7 @@
     "nextcloud-server": "^0.15.9",
     "nextcloud-vue": "^0.11.4",
     "nextcloud_issuetemplate_builder": "^0.1.0",
+    "printscout": "^2.0.3",
```
Comment (ChristophWurst) [inline package.json]: new dep is fine :+1: but please use `"2.0.3"` instead so another `npm i` does not [accidentally update this package](https://stackoverflow.com/questions/45022048/why-does-npm-install-rewrite-package-lock-json)

File: templates/index.php [hunk 2]
```diff
@@ -24,6 +24,7 @@
  *
  */
 style('mail', 'mail');
+style('mail', 'print');
```
Comment (ChristophWurst) [inline templates/index.php]: any chance we can inline these styles to the Vue components? We are trying to get the number of requests down in general

Comment (sourenaraya) [reply templates/index.php]: ok, moved styles to `Message.vue`

File: package.json (hunk 1 above)
Comment (sourenaraya) [reply package.json]: fixed, thanks

**Comment by ChristophWurst** [comment]
Let's wait for @jancborchardt's review

**Comment by sourenaraya** [comment]
Sorry for asking this again, but, any update on this?

**Comment by jancborchardt** [comment]
Sorry, I was recently on vacation and have to get up to speed again. :) Will check it soon. @sourenaraya just for clarification: There’s no interface change to check for, only Ctrl-P and browser menu print functionality, correct?

**Comment by sourenaraya** [comment]
Yep, just Ctrl+P as usual.

**Comment by jancborchardt** [comment]
Works very nicely, good job @sourenaraya! Only details:
- In Firefox: If you print via the menu → Print functionality you get a preview. If you scroll there, the `mail-message-header` (with subject, sender and recipient) is still sticky so floating above the text. Maybe it should be changed to `position: relative;` for printing? It does come out fine in the PDF
- Also in Firefox/Ubuntu, when printing to PDF, is it possible to preset the file name to something, like the mail title? For me the filename is always the same as last time unless I rename.
- In the printout, the date of the email is not included. Should we have this? Might be something for a follow-up too.

In general nothing blocking and we could also merge it like this :+1:  :)

**Comment by sourenaraya** [comment]
i've set `position: relative` to `mail-message-header`, thanks.

* print-to-pdf function: AFAK there is no way to suggest filename to browser. Its always `mozilla.pdf`, or something set in `about:config`
* date in the printout:  date is not present on plaintext, as well as on richtext emails, anywhere. First of all, we have to somehow c it in backend responses, or have a way to retrieve date information from backend. If there is any, i am not aware of it.

**Comment by ChristophWurst** [comment]
@sourenaraya please don't forget to go to https://www.bountysource.com/issues/50806693-print-stylesheet and claim your bounty :)

**Comment by sourenaraya** [comment]
Issue still marked as open on the bountysource 😂

**Comment by ChristophWurst** [comment]
Ugh. Bountysource is so buggy. Sorry about that. Tried to get support via https://twitter.com/ChristophWurst/status/1159344197436104704. Let's see :pray:

**Pull request merged**